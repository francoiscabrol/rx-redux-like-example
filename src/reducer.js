
import * as actions from './actions';

// Redux reducer
const reducer = (state, action) => {
  switch (action.type) {
    case actions.UPDATE_NAME_START: {
      return {
        ...state,
        loading: true
      };
    }
    case actions.UPDATE_NAME_FAILURE: {
      return {
        ...state,
        loading: false,
        error: action.error
      };
    }
    case actions.UPDATE_NAME_SUCCESS: {
      return {
        ...state,
        loading: false,
        name: action.name
      };
    }
    default: {
      return state;
    }
  }
};

export default reducer;
