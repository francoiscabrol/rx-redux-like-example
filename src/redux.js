import React from 'react';
import Rx from 'rxjs/Rx';
import PropTypes from 'prop-types';
import R from 'ramda';

export class Provider extends React.Component {
  getChildContext() {
    return {
      store: this.props.store
    };
  }
  render() {
    return this.props.children;
  }
}

Provider.childContextTypes = {
  store: PropTypes.object
};

export function connect(mapStateToProps = () => ({}), mapDispatchToProps = () => ({})) {
  return Component => {
    class Connected extends React.Component {
      onStoreOrPropsChange(props, stateProps = {}) {
        const {store} = this.context;
        const dispatchProps = mapDispatchToProps((action) => store.dispatch(action), props);
        this.setState({
          ...stateProps,
          ...dispatchProps
        });
      }
      componentWillMount() {
        const {store} = this.context;
        const statePropsObservable = store.store$.map(state => mapStateToProps(state, this.props)).filter(stateProps => {
          if (this.state === null) return true
          const currentState = R.pick(Object.keys(stateProps), this.state);
          return !R.equals(currentState, stateProps)
        });

        this.unsubscribe = statePropsObservable
          .subscribe((stateProps) => {
            console.log("Re-render the connected component with new props:", stateProps);
            return this.onStoreOrPropsChange(this.props, stateProps);
          });
      }
      componentWillUnmount() {
        this.unsubscribe();
      }
      render() {
        return <Component {...this.props} {...this.state}/>;
      }
    }

    Connected.contextTypes = {
      store: PropTypes.object
    };

    return Connected;
  };
};

export class Store {

  constructor(initState, reducer, middleware) {
    const withDevTools = typeof window !== 'undefined' && window.devToolsExtension;

    if (withDevTools)
      this.devTools = window.devToolsExtension.connect();

    // create our stream as a subject so arbitrary data can be sent on the stream
    this.action$ = new Rx.Subject();

    // Reduxification
    this.store$ = this.action$
      .startWith(initState)
      .scan((state, action) => {
        const newState = reducer(state, action);
        if (withDevTools)
          this.devTools.send(action.type, newState);
        return newState;
      });

    this.middleware = middleware;

    this.coreDispatch = (action) => {
      this.action$.next(action);
    }

    if (this.middleware) {
      const dispatch = action => this.dispatch(action);
      const cancel = (cancelType) => this.action$.filter(({ type }) => type === cancelType);
      this.dispatch = this.middleware({dispatch, getState: this.getState, cancel})(this.coreDispatch);
    } else {
      this.dispatch = this.coreDispatch;
    }
  }

  getState() {
    return this.store$; // should be synchronous
  }

  subscribe(rendererCallback) {
    this.store$.subscribe(rendererCallback);
  }
}

export const thunkMiddleware = ({dispatch, getState, cancel}) => next => action => {
  if (typeof action === 'function') {
    return action({dispatch, getState, cancel});
  }
  return next(action);
};
