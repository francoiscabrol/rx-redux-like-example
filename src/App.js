// @flow

import React from "react";
import { connect } from "./redux";
import { updateName, cancelUpdateName } from "./actionCreators";

// React view component
const DynamicName = props => {
  const { name, subtitle, onChangeName, onCancelChangeName } = props;
  return (
    <div>
      <h1>
        {name}
      </h1>
      <h2>
        {subtitle}
      </h2>
      <button onClick={() => onChangeName("Harry")}>Harry</button>
      <button onClick={() => onChangeName("Sally")}>Sally</button>
      <button onClick={() => onCancelChangeName()}>Cancel</button>
    </div>
  );
};

const ConnectedApp = connect(
  state => {
    return {
      name: state.name
    };
  },
  dispatch => {
    return {
      onChangeName: name => dispatch(updateName(name)),
      onCancelChangeName: () => dispatch(cancelUpdateName())
    };
  }
)(DynamicName);

export default ConnectedApp;
