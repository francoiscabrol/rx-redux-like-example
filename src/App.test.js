import React from "react";
import { Provider, Store, applyMiddleware, thunkMiddleware } from "./redux";
import renderer from 'react-test-renderer';
import reducer from "./reducers";
import App from "./App";

const message  = 'you never clicked'
const store = new Store({ message: {value: message } }, reducer, thunkMiddleware);

describe('Welcome', () => {
  it('Welcome renders hello world', () => {
    const app = renderer.create(<Provider store={store}>
      <App subtitle="Un good test" />
    </Provider>);
    expect(app.toJSON()).toMatchSnapshot()
  });
});
