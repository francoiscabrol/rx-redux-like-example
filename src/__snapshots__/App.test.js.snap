// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Welcome Welcome renders hello world 1`] = `
<div>
  <h1>
    NoName
  </h1>
  <h2>
    Un good test
  </h2>
  <h3>
    you never clicked
  </h3>
  <button
    onClick={[Function]}
  >
    Click test
  </button>
  <button
    onClick={[Function]}
  >
    Harry
  </button>
  <button
    onClick={[Function]}
  >
    Sally
  </button>
  <button
    onClick={[Function]}
  >
    Cancel
  </button>
</div>
`;
